File: typescript/src/fold.ts
import {listService} from "./listService";
import {Result, success} from "for-comprehension-ts";

const results: Result<number[], string>[] = [2, -1, 4].map(listService)

const zero: Result<number[], string> = success([])

const combination: (collector: Result<number[], string>, next: Result<number[], string>) => Result<number[], string> =
    (collector, next) => collector.flatMap(cList => next.map(nList => cList.concat(nList)))

const combined: Result<number[], string> = results.reduce(combination, zero)

console.log(combined)

File: swift/Fold.swift

func fold() {
    printName()
    
    let results = [2, -1, 4].map(listService)
    
    let zero: Result<[Int], String> = .success([])

    let combined = results.reduce(zero) { collector, next in
        collector.flatMap { cList in
            next.map { cList + $0 }
        }
    }

    print(combined)
}
